crates/autopilot/src/domain/settlement/trade/mod.rs

```diff
@@ -0,0 +1,218 @@
+    /// before and after applying the fees.
+    pub fn fee_in_sell_token(&self) -> Result<eth::SellTokenAmount, math::Error> {
+        self.as_math().fee_in_sell_token()
+    }
+
+    /// Protocol fees are defined by fee policies attached to the order.
+    pub fn protocol_fees_in_sell_token(
+        &self,
+        auction: &super::Auction,
+    ) -> Result<Vec<(eth::SellTokenAmount, fee::Policy)>, math::Error> {
+        self.as_math().protocol_fees_in_sell_token(auction)
+    }
+
+    /// Represent the current trade as a math trade, for which we know how
+    /// to calculate surplus, fees, etc.
+    fn as_math(&self) -> math::Trade {
+        match self {
+            Trade::Fulfillment(trade) => trade.as_math(),
+            Trade::Jit(trade) => trade.as_math(),
+        }
+    }
+
+    pub fn new(
+        trade: transaction::EncodedTrade,
+        auction: &super::Auction,
+        database_orders: &HashSet<domain::OrderUid>,
+        created: u32,
+    ) -> Self {
+        // If it's not in a database, then it's definitely a JIT order
+        if !database_orders.contains(&trade.uid) {
+            Trade::Jit(Jit {
+                uid: trade.uid,
+                sell: trade.sell,
+                buy: trade.buy,
+                side: trade.side,
+                receiver: trade.receiver,
+                valid_to: trade.valid_to,
+                app_data: trade.app_data,
+                fee_amount: trade.fee_amount,
+                sell_token_balance: trade.sell_token_balance,
+                buy_token_balance: trade.buy_token_balance,
+                signature: trade.signature,
+                executed: trade.executed,
+                prices: trade.prices,
+                created,
+                surplus_capturing: auction
+                    .surplus_capturing_jit_order_owners
+                    .contains(&trade.uid.owner()),
+            })
+        }
+        // If it's in a database, then it's a normal fulfillment of an user order
+        else {
+            Trade::Fulfillment(Fulfillment {
+                uid: trade.uid,
+                sell: trade.sell,
+                buy: trade.buy,
+                side: trade.side,
+                executed: trade.executed,
+                prices: trade.prices,
+                in_auction: auction.orders.contains_key(&trade.uid),
+            })
+        }
+    }
+}
+
+/// Trade representing an user trade. User trades are part of the orderbook.
+#[derive(Debug, Clone)]
+pub struct Fulfillment {
+    pub uid: domain::OrderUid,
+    pub sell: eth::Asset,
+    pub buy: eth::Asset,
+    pub side: order::Side,
+    pub executed: order::TargetAmount,
+    pub prices: Prices,
+    // Whether the trade was part of an auction
+    pub in_auction: bool,
+}
+
+impl Fulfillment {
+    /// Converts the trade to a math trade, which can be used to calculate
+    /// surplus, fees, etc.
+    pub fn as_math(&self) -> math::Trade {
```

Comment:
`as_math()` looks pretty strange at the call site.
Also `as_somehing()` usually indicates a conversion that is "free" and operates with references. This would be more like a `to_math()`.

Maybe it would be slightly nicer to read if the callsite would be sth like:
```rust
compute::fees_in_ether(&self);
```
Which could be something like:
```rust
mod compute {
    struct ExecutionDetails {
        // all the "math" fields
    }

    fn fees_in_ether(impl Into<ExecutionDetails>, extra_data: ()); 
}
```

not 100% sure this would work with the visibility rules but if it does I think this could be nicer to read.

Reply:
Agree, this `as_math` also trigggered me. I think as long as we keep math::Trade completely private to the rest of the domain it's ok to keep without thinking too much of what this actually represents, but these methods are public (they should only be callable from Fulfillment).

I wonder if we can't simply have `math::Trade::from_fullfilment` and `math::Trade::form_jit`

Reply:
refactored